x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/writer/JsonDataToProcessWriter.java

```diff
@@ -160,6 +160,11 @@ private void writeJson(CategorizationAnalyzer categorizationAnalyzer, XContentPa
             if (categorizationAnalyzer != null && categorizationFieldIndex != null) {
                 tokenizeForCategorization(categorizationAnalyzer, input[categorizationFieldIndex], record);
             }
+
+            if (categorizationFieldIndex != null) {
+                record[categorizationFieldIndex] = analysisConfig.maybeTruncateCatgeorizationField(record[categorizationFieldIndex]);
```

Comment:
This is assuming indexes into the input array are identical to indexes into the output array.  I don't think that's always the case, otherwise lines 156-157 are pointless.  So I think we'll need to separately remember the index into the output array or else do the truncation at a different point.  Since we tokenise the value from the input array we could do the truncation in the loop around lines 156/157 so the output array contains a different string to the input array for this value.